File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/domain/trans/SaleTrans.java
package com.bits.bee.bpmc.domain.trans;

import java.lang.System;

/**
 * Created by aldi on 19/05/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000V\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\b\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000b\n\u0002\b\n\n\u0002\u0010\u000e\n\u0002\b\u0006\n\u0002\u0010 \n\u0002\b\u000f\b\u0007\u0018\u00002\u000e\u0012\u0004\u0012\u00020\u0002\u0012\u0004\u0012\u00020\u00030\u0001B\u0007\b\u0007\u00a2\u0006\u0002\u0010\u0004J\"\u0010\u0014\u001a\u00020\u00152\u0006\u0010\u0016\u001a\u00020\u00172\b\b\u0002\u0010\u0018\u001a\u00020\u00192\b\b\u0002\u0010\u001a\u001a\u00020\u0019J\u0010\u0010\u0014\u001a\u00020\u00152\u0006\u0010\u001b\u001a\u00020\u0003H\u0016J\u0006\u0010\u001c\u001a\u00020\u0015J\b\u0010\u001d\u001a\u00020\u0015H\u0002J\u000e\u0010\u001e\u001a\u00020\u00152\u0006\u0010\u001f\u001a\u00020\u0003J\u0010\u0010 \u001a\u00020\u00152\u0006\u0010!\u001a\u00020\u0003H\u0002J\u000e\u0010\"\u001a\u00020\u00152\u0006\u0010#\u001a\u00020$J\u000e\u0010%\u001a\u00020\u00152\u0006\u0010&\u001a\u00020\u0003J\u0018\u0010\'\u001a\u00020\u00032\u0006\u0010!\u001a\u00020\u00032\u0006\u0010(\u001a\u00020\u0003H\u0002J\b\u0010)\u001a\u00020\u0002H\u0016J\u0016\u0010*\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010\u00030+2\u0006\u0010,\u001a\u00020\u0003J\u0010\u0010-\u001a\u00020\u00152\u0006\u0010.\u001a\u00020\u0002H\u0016J\u0006\u0010/\u001a\u00020\u0015J\u0006\u00100\u001a\u00020\u0015J\u0016\u00101\u001a\u00020\u00152\u0006\u00102\u001a\u00020\u00032\u0006\u00103\u001a\u00020\u0003J\b\u00104\u001a\u00020\u0015H\u0016J\b\u00105\u001a\u00020\u0015H\u0016J\u000e\u00106\u001a\u00020\u00152\u0006\u0010\u000b\u001a\u00020\fJ\u0010\u00107\u001a\u00020\u00152\b\u00108\u001a\u0004\u0018\u00010\u0011J\b\u00109\u001a\u00020\u0015H\u0016R\u001c\u0010\u0005\u001a\u0004\u0018\u00010\u0006X\u0086\u000e\u00a2\u0006\u000e\n\u0000\u001a\u0004\b\u0007\u0010\b\"\u0004\b\t\u0010\nR\u0010\u0010\u000b\u001a\u0004\u0018\u00010\fX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u0010\u0010\r\u001a\u0004\u0018\u00010\u0003X\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u000e\u001a\u00020\u000fX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u0010\u0010\u0010\u001a\u0004\u0018\u00010\u0011X\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0012\u001a\u00020\u000fX\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u0010\u0010\u0013\u001a\u0004\u0018\u00010\u0003X\u0082\u000e\u00a2\u0006\u0002\n\u0000\u00a8\u0006:"}, d2 = {"Lcom/bits/bee/bpmc/domain/trans/SaleTrans;", "Lcom/bits/bee/bpmc/domain/trans/BaseTrans;", "Lcom/bits/bee/bpmc/domain/model/Sale;", "Lcom/bits/bee/bpmc/domain/model/Saled;", "()V", "addOnTrans", "Lcom/bits/bee/bpmc/domain/trans/AddOnTrans;", "getAddOnTrans", "()Lcom/bits/bee/bpmc/domain/trans/AddOnTrans;", "setAddOnTrans", "(Lcom/bits/bee/bpmc/domain/trans/AddOnTrans;)V", "bp", "Lcom/bits/bee/bpmc/domain/model/Bp;", "currentSaled", "dnoCounter", "", "grpAddon", "Lcom/bits/bee/bpmc/domain/model/ItemGroup;", "roundVal", "saledParent", "addDetail", "", "itemWithUnit", "Lcom/bits/bee/bpmc/domain/model/ItemWithUnit;", "isBonus", "", "useItemqty", "d", "calcDetailDiskon", "calculate", "deleteDetail", "saledDel", "deleteOldSaleAddonD", "saled", "diskonMaster", "diskon", "", "editDetail", "saledEdit", "editSaled", "newSaled", "getMaster", "getSaledByUpSaledList", "", "model", "loadTrans", "var1", "mergeAddon", "mergeItemAddon", "mergeSaled", "saled1", "saled2", "newTrans", "saveTrans", "setBp", "setGrpAddOn", "itemGroup", "voidTrans", "app_debug"})
@javax.inject.Singleton()
public final class SaleTrans extends com.bits.bee.bpmc.domain.trans.BaseTrans<com.bits.bee.bpmc.domain.model.Sale, com.bits.bee.bpmc.domain.model.Saled> {
    private com.bits.bee.bpmc.domain.model.Bp bp;
    private com.bits.bee.bpmc.domain.model.Saled saledParent;
    private com.bits.bee.bpmc.domain.model.Saled currentSaled;
    private com.bits.bee.bpmc.domain.model.ItemGroup grpAddon;
    private int dnoCounter = 0;
    private int roundVal = 0;
    @org.jetbrains.annotations.Nullable()
    private com.bits.bee.bpmc.domain.trans.AddOnTrans addOnTrans;
    
    @javax.inject.Inject()
    public SaleTrans() {
        super();
    }
    
    @org.jetbrains.annotations.Nullable()
    public final com.bits.bee.bpmc.domain.trans.AddOnTrans getAddOnTrans() {
        return null;
    }
    
    public final void setAddOnTrans(@org.jetbrains.annotations.Nullable()
    com.bits.bee.bpmc.domain.trans.AddOnTrans p0) {
    }
    
    public final void setBp(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Bp bp) {
    }
    
    public final void setGrpAddOn(@org.jetbrains.annotations.Nullable()
    com.bits.bee.bpmc.domain.model.ItemGroup itemGroup) {
    }
    
    @java.lang.Override()
    public void newTrans() {
    }
    
    @java.lang.Override()
    public void saveTrans() {
    }
    
    @java.lang.Override()
    public void voidTrans() {
    }
    
    @java.lang.Override()
    public void loadTrans(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Sale var1) {
    }
    
    public final void addDetail(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.ItemWithUnit itemWithUnit, boolean isBonus, boolean useItemqty) {
    }
    
    @java.lang.Override()
    public void addDetail(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Saled d) {
    }
    
    public final void editDetail(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Saled saledEdit) {
    }
    
    public final void diskonMaster(@org.jetbrains.annotations.NotNull()
    java.lang.String diskon) {
    }
    
    public final void calcDetailDiskon() {
    }
    
    private final com.bits.bee.bpmc.domain.model.Saled editSaled(com.bits.bee.bpmc.domain.model.Saled saled, com.bits.bee.bpmc.domain.model.Saled newSaled) {
        return null;
    }
    
    public final void deleteDetail(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Saled saledDel) {
    }
    
    private final void calculate() {
    }
    
    public final void mergeAddon() {
    }
    
    public final void mergeItemAddon() {
    }
    
    public final void mergeSaled(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Saled saled1, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Saled saled2) {
    }
    
    private final void deleteOldSaleAddonD(com.bits.bee.bpmc.domain.model.Saled saled) {
    }
    
    @org.jetbrains.annotations.NotNull()
    public final java.util.List<com.bits.bee.bpmc.domain.model.Saled> getSaledByUpSaledList(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.model.Saled model) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public com.bits.bee.bpmc.domain.model.Sale getMaster() {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/domain/trans/SaleTrans.kt
package com.bits.bee.bpmc.domain.trans

import com.bits.bee.bpmc.data.data_source.local.model.PromoEntity
import com.bits.bee.bpmc.data.list.ListPromoBonus
import com.bits.bee.bpmc.domain.calc.PromoCalc
import com.bits.bee.bpmc.domain.calc.SaleCalc
import com.bits.bee.bpmc.domain.model.*
import com.bits.bee.bpmc.domain.repository.PromoRepository
import com.bits.bee.bpmc.utils.BPMConstants
import com.bits.bee.bpmc.utils.CalcUtils
import kotlinx.coroutines.flow.first
import java.math.BigDecimal
import java.math.RoundingMode
import javax.inject.Inject
import javax.inject.Singleton


/**
 * Created by aldi on 19/05/22.
 */
@Singleton
class SaleTrans @Inject constructor(
    private val promoCalc: PromoCalc,
    private val saleCalc: SaleCalc,
    private val promoRepository: PromoRepository
) : BaseTrans<Sale, Saled>() {

    private var bp: Bp? = null
    private var saledParent: Saled? = null
    private var currentSaled: Saled? = null
    private var grpAddon : ItemGroup? = null
    var salePromoList: MutableList<SalePromo> = mutableListOf()
    private var dnoCounter = 0
    private var roundVal = 0
    var addOnTrans: AddOnTrans? = null

    suspend fun setBp(bp: Bp){
        this.bp = bp
        getMaster().bpId = bp.id!!
        getMaster().bp = bp
        getMaster().bpName = bp.name
        calculate()
    }

    fun setGrpAddOn(itemGroup: ItemGroup?){
        this.grpAddon = itemGroup
    }

    override fun newTrans() {
        mTblMaster = Sale()
        mTblDetail.clear()
        salePromoList.clear()
        promoCalc.listPromoBonus.clear()
        addOnTrans = null
    }

    override fun saveTrans() {

    }

    override fun voidTrans() {
    }

    override fun loadTrans(var1: Sale) {

    }

    suspend fun loadTrans(sale : Sale, saledList : MutableList<Saled>, saleAddOn: SaleAddOn? = null,
                  saleAddOnDList : MutableList<SaleAddOnD> = mutableListOf(),
                  salePromoList: MutableList<SalePromo> = mutableListOf()){
        mTblMaster = sale
        mTblDetail = saledList
        if(saleAddOn != null){
            if(addOnTrans == null)
                addOnTrans = AddOnTrans()

            addOnTrans!!.loadTrans(saleAddOn, saleAddOnDList)
            addOnTrans!!.getListDetail().forEach { saleAddOnD ->
                getListDetail().forEach { saled ->
                    if(saleAddOnD.upSaled?.id == saled.id){
                        saleAddOnD.upSaled = saled
                    }
                    if(saleAddOnD.saled?.id == saled.id){
                        saleAddOnD.saled = saled
                    }
                }
            }
        }
        this.salePromoList = salePromoList
       this.salePromoList.forEach { salePromo ->
           getListDetail().forEach { saled ->
               if(salePromo.saled?.id == saled.id){
                   val promo = promoRepository.getPromoById(salePromo.promo!!.id).first()
                   if(promo.promoCat == PromoEntity.BONUS && salePromo.promoRule == "D") {
                       saled.isBonus = true
                       saled.isBonusUsed = true
                   }
                   salePromo.saled = saled
               }
           }
        }
    }


    suspend fun addDetail(itemWithUnit: ItemWithUnit, isBonus: Boolean = false, useItemqty: Boolean = false, promoBonus: ListPromoBonus.PromoBonus? = null) {
        val item = itemWithUnit.item
        val discExp = itemWithUnit.discExp
        val discAmt = itemWithUnit.discAmt
        val unit = itemWithUnit.unit
        val pid = itemWithUnit.pid
        val stock = itemWithUnit.stock

        if (getListDetail().isEmpty()) {
            /** when list empty add new detail to list and set current saled*/
            val saledNew = Saled(
                itemId = item.id,
                itemCode = item.code,
                listPrice = item.price,
                qty = if(useItemqty) item.qty else BigDecimal.ONE,
                name = item.name1,
                discExp = discExp,
                discAmt = discAmt,
                disc2Amt = BigDecimal.ZERO,
                isBonus = isBonus,
                isBonusUsed = isBonus,
                tax = if(item.tax.isEmpty()) BigDecimal.ZERO else BigDecimal(item.tax),
                crcId = item.crcId,
                crcSymbol = item.crcSymbol,
                unitId = unit?.id,
                unit = unit?.unit,
                conv = unit?.conv,
                pid = pid,
                isAddOn = item.isAddOn,
                item = item,
                taxCode = item.taxCode,
                stock = stock
            )
            addDetail(saledNew)
            currentSaled = saledNew
        } else {
            var isNew = true
            for (saled in getListDetail()) {
                if (saled.itemId == item.id && !item.isVariant && !item.isHaveAddOn && (grpAddon == null || (grpAddon != null && item.itemGrpId != grpAddon!!.id))) {
                    saled.listPrice = item.price
                    saled.qty = saled.qty.add(if (useItemqty) item.qty else BigDecimal.ONE)
                    if (saled.disc != BigDecimal.ZERO) {
                        if (!saled.discExp.contains("%")) {
                            saled.disc = saled.disc
                            saled.discAmt = saled.disc.setScale(roundVal, RoundingMode.HALF_UP)
                            saled.discExp = saled.discExp
                        } else if (saled.discExp.contains("%")) {
                            saled.disc = saled.disc
                            saled.discAmt =
                                saled.listPrice.multiply(saled.disc, BPMConstants.MC_FOUR)
                                    .divide(BigDecimal(100), BPMConstants.MC_FOUR)
                                    .setScale(roundVal, RoundingMode.HALF_UP)

                            saled.discExp = saled.discExp
                        }
                    }
                    currentSaled = saled
                    isNew = false
                    calculate()
                    break
                }
            }
            if(isNew) {
                val saledNew = Saled(
                    itemId = item.id,
                    itemCode = item.code,
                    listPrice = item.price,
                    qty = if (useItemqty) item.qty else BigDecimal.ONE,
                    name = item.name1,
                    discExp = "0",
                    disc2Amt = BigDecimal.ZERO,
                    isBonus = isBonus,
                    isBonusUsed = isBonus,
                    tax = if (item.tax.isEmpty()) BigDecimal.ZERO else BigDecimal(item.tax),
                    crcId = item.crcId,
                    crcSymbol = item.crcSymbol,
                    unitId = unit?.id,
                    unit = unit?.unit,
                    conv = unit?.conv,
                    pid = pid,
                    isAddOn = item.isAddOn,
                    item = item,
                    taxCode = item.taxCode,
                    stock = stock
                )
                addDetail(saledNew)
                currentSaled = saledNew
            }
        }
        if (grpAddon != null && item.itemGrpId == grpAddon!!.id) {
            if (addOnTrans == null) {
                addOnTrans = AddOnTrans()
                addOnTrans!!.newTrans()
                addOnTrans!!.getMaster().saleId = getMaster()
            }
            val saleAddOnD = SaleAddOnD()
            saleAddOnD.saleAddOn = addOnTrans?.getMaster()
            saleAddOnD.saled = currentSaled
            saleAddOnD.upSaled = saledParent
            addOnTrans?.addDetail(saleAddOnD)
        } else {
            this.saledParent = currentSaled
        }
        if (isBonus) {
//        Tambah salebonus apabila item bonus
            promoBonus?.let {
                promoCalc.addSalePromoItem(promoBonus.getPromo(), promoBonus.saled, currentSaled)
            }
        }
    }


    suspend fun addDetail(itemWithUnit: ItemWithUnit, saledParent : Saled){
        val item = itemWithUnit.item
        val discExp = itemWithUnit.discExp
        val discAmt = itemWithUnit.discAmt
        val unit = itemWithUnit.unit
        val pid = itemWithUnit.pid

        if (getListDetail().isEmpty()) {
            /** when list empty add new detail to list and set current saled*/
            val saledNew = Saled(
                itemId = item.id,
                itemCode = item.code,
                listPrice = item.price,
                qty = if(grpAddon != null && item.itemGrpId == grpAddon!!.id) item.qty else BigDecimal.ONE,
                name = item.name1,
                discExp = discExp,
                discAmt = discAmt,
                disc2Amt = BigDecimal.ZERO,
                tax = if(item.tax.isEmpty()) BigDecimal.ZERO else BigDecimal(item.tax),
                crcId = item.crcId,
                crcSymbol = item.crcSymbol,
                unitId = unit?.id,
                unit = unit?.unit,
                conv = unit?.conv,
                pid = pid,
                isAddOn = item.isAddOn,
                item = item
            )
            addDetail(saledNew)
            currentSaled = saledNew
        } else {
            var isNew = true
            for (saled in getListDetail()) {
                if (saled.itemId == item.id && !item.isVariant && !item.isHaveAddOn && (grpAddon == null || (grpAddon != null && item.itemGrpId == grpAddon!!.id))) {
                    saled.listPrice = item.price
                    saled.qty = saled.qty.add(BigDecimal.ONE)
                    if (saled.disc != BigDecimal.ZERO) {
                        if (!saled.discExp.contains("%")) {
                            saled.disc = saled.disc
                            saled.discAmt = saled.disc.setScale(roundVal, RoundingMode.HALF_UP)
                            saled.discExp = saled.discExp
                        } else if (saled.discExp.contains("%")) {
                            saled.disc = saled.disc
                            saled.discAmt =
                                saled.listPrice.multiply(saled.disc, BPMConstants.MC_FOUR)
                                    .divide(BigDecimal(100), BPMConstants.MC_FOUR)
                                    .setScale(roundVal, RoundingMode.HALF_UP)

                            saled.discExp = saled.discExp
                        }
                    }
                    currentSaled = saled
                    isNew = false
                    calculate()
                    break
                }
            }
            if(isNew) {
                val saledNew = Saled(
                    itemId = item.id,
                    itemCode = item.code,
                    listPrice = item.price,
                    qty = if(grpAddon != null && item.itemGrpId == grpAddon!!.id) item.qty else BigDecimal.ONE,
                    name = item.name1,
                    discExp = "0",
                    disc2Amt = BigDecimal.ZERO,
                    tax = if (item.tax.isEmpty()) BigDecimal.ZERO else BigDecimal(item.tax),
                    crcId = item.crcId,
                    crcSymbol = item.crcSymbol,
                    unitId = unit?.id,
                    unit = unit?.unit,
                    conv = unit?.conv,
                    pid = pid,
                    isAddOn = item.isAddOn,
                    item = item
                )
                addDetail(saledNew)
                currentSaled = saledNew
            }
        }
        if (grpAddon != null && item.itemGrpId == grpAddon!!.id) {
            if (addOnTrans == null) {
                addOnTrans = AddOnTrans()
                addOnTrans!!.newTrans()
                addOnTrans!!.getMaster().saleId = getMaster()
            }
            val saleAddOnD = SaleAddOnD()
            saleAddOnD.saleAddOn = addOnTrans?.getMaster()
            saleAddOnD.saled = currentSaled
            saleAddOnD.upSaled = saledParent
            addOnTrans?.addDetail(saleAddOnD)
        } else {
            this.saledParent = currentSaled
        }
    }

    override suspend fun addDetail(d: Saled) {
        dnoCounter++
        d.dno = dnoCounter
        super.addDetail(d)
        calculate()
    }


    suspend fun editDetail(saledEdit: Saled) {
        getListDetail().map { saled ->
            if(saled === saledEdit) {
//                if(saled.item != null && saled.item!!.isHaveAddOn){
//                    addOnTrans?.getListDetail()?.forEach { saleAddOnD ->
//                        if(saleAddOnD.upSaled == saled){
//                            getListDetail().forEach { saled1 ->
//                                val saled2 = saleAddOnD.saled
//                                if(saled1 === saled2){
//                                    val qtyPcs = saled2.qty / saledEdit.qty
//                                    saled1.qty = qtyPcs.multiply(newQty)
//                                    editSaled(saled1, saled2)
//                                }
//                            }
//                        }
//                    }
//                }
                editSaled(saled, saledEdit)
            }
        }
        calculate()
    }

    suspend fun updateDiskonMaster(diskon : String)  {
        val sale = getMaster()
        if(diskon.isNotEmpty()) {
            sale.discExp = diskon
            sale.discAmt = CalcUtils.getDiscAmt(diskon, sale.subtotal)
        } else {
            sale.discExp = ""
            sale.discAmt = BigDecimal.ZERO
        }
        calcDetailDiskon()
        calculate()
    }

    private suspend fun calcDetailDiskon() {
        val sale = getMaster()
        for (saled in getListDetail()){
            var disc2Amt = BigDecimal.ZERO
            if(sale.discAmt > BigDecimal.ZERO){
                disc2Amt = sale.discAmt.multiply(saled.subtotal).divide(sale.subtotal, BPMConstants.MC_FOUR).divide(saled.qty, BPMConstants.MC_FOUR)
            }
            saled.disc2Amt = disc2Amt
        }
        calculate()
    }

    private fun editSaled(saled: Saled, newSaled: Saled) : Saled  {
        saled.listPrice =  newSaled.listPrice
        saled.qty = newSaled.qty
        saled.disc = newSaled.disc
        saled.discExp = newSaled.discExp
        saled.discAmt = newSaled.discAmt
        saled.tax = newSaled.tax
        saled.taxAmt = newSaled.taxAmt
        saled.disc2Amt = newSaled.disc2Amt
        saled.dNotes = newSaled.dNotes
        return saled
    }

    suspend fun deleteDetail(saledDel: Saled) {
        var indexDelete = -1

        for (i in 0 until getListDetail().size){
            val saled = getListDetail()[i]
            if(saledDel === saled){
                indexDelete = i
                break
            }
        }
        if(indexDelete > -1) {
            getListDetail().removeAt(indexDelete)
        }
        calculate()
    }

    suspend fun calculate(){
        saleCalc.calculate(getMaster(), getListDetail(), bp!!)
    }

    suspend fun mergeAddon() {
        //Rubah data saleaddon ke map
        val mapAddOn: MutableMap<Saled, MutableList<Saled>> = HashMap()
        //Map untuk seleksi  addon yg dimerge
        val mapSaled: MutableMap<Saled, List<Saled>> = HashMap()
        if (addOnTrans != null) {
            for (saleAddOnD in addOnTrans!!.getListDetail()) {
                val up_saled: Saled? = saleAddOnD.upSaled
                val saled: Saled? = saleAddOnD.saled
                if (mapAddOn.containsKey(up_saled)) {
                    if(saled != null)
                        mapAddOn[up_saled]!!.add(saled)
                } else {
                    if(up_saled != null && saled != null)
                        mapAddOn[up_saled] = ArrayList(listOf(saled))
                }
            }
            for ((addOnKey, valueAddOn) in mapAddOn) {
                if (mapSaled.isNotEmpty()) {
                    var exist = false
                    //matching saled dan addon, apabila reorder qty dipecah menjadi per satuan
                    for ((saledKey, valueSaled) in mapSaled) {
                        if (addOnKey.itemId == saledKey.itemId && addOnKey.listPrice.compareTo(saledKey.listPrice) == 0) {
                            val addonItem: MutableList<String> = ArrayList()
                            val saledItem: MutableList<String> = ArrayList()
                            for (saled in valueAddOn) {
                                addonItem.add(saled.name + saled.qty.divide(addOnKey.qty))
                            }
                            for (saled in valueSaled) {
                                saledItem.add(saled.name + saled.qty.divide(saledKey.qty))
                            }
                            addonItem.sort()
                            saledItem.sort()
                            if (addonItem == saledItem) {
                                mergeSaled(saledKey, addOnKey)
                                for (saled in valueSaled) {
                                    for (saled1 in valueAddOn) {
                                        if (saled.itemId == saled1.itemId) {
                                            mergeSaled(saled, saled1)
                                            deleteOldSaleAddonD(addOnKey)
                                        }
                                    }
                                }
                                exist = true
                            }
                        }
                    }
                    if (!exist) {
                        mapSaled[addOnKey] = valueAddOn
                    }
                } else {
                    mapSaled[addOnKey] = valueAddOn
                }
            }
        }
    }

    //Merge item addon tanpa addon
    suspend fun mergeItemAddon() {
        if (addOnTrans != null) {
            val saledList: MutableList<Saled> = ArrayList()
            val saledListResult: MutableList<Saled> = ArrayList()
            for (saled in getListDetail()) {
                if (saled.item!!.isHaveAddOn) {
                    var isMerge = true
                    for (saleAddOnD in addOnTrans!!.getListDetail()) {
                        if (saled == saleAddOnD.upSaled) {
                            isMerge = false
                            break
                        }
                    }

                    if (isMerge) {
                        saledList.add(saled)
                    }
                }
            }
            if (saledList.size > 0) {
                for (saled in saledList) {
                    var isExist = false
                    if (saledListResult.size == 0) {
                        saledListResult.add(saled)
                    } else {
                        for (saled1 in saledListResult) {
                            if (saled.itemId == saled1.itemId && saled.listPrice.compareTo(saled1.listPrice) == 0) {
                                mergeSaled(saled1, saled)
                                isExist = true
                                break
                            }
                        }
                        if (!isExist) {
                            saledListResult.add(saled)
                        }
                    }
                }
            }
        }
    }

    suspend fun mergeSaled(saled1: Saled, saled2: Saled) {
        for (saled in getListDetail()) {
            if (saled === saled1) {
                saled.listPrice = saled.item?.price ?: BigDecimal.ZERO
                saled.qty = saled.qty.add(saled2.qty)
                if (saled.disc != BigDecimal.ZERO) {
                    if (!saled.discExp.contains("%")) {
                        saled.disc = saled.disc
                        saled.discAmt = saled.disc.setScale(roundVal, RoundingMode.HALF_UP)
                        saled.discExp = saled.discExp
                    } else if (saled.discExp.contains("%")) {
                        saled.disc = saled.disc
                        saled.discAmt = (saled.listPrice * saled.disc / BigDecimal(100)).setScale(roundVal, RoundingMode.HALF_UP)
                        saled.discExp = saled.discExp
                    }
                }
                calculate()
            }
        }
        for (saled in getListDetail()) {
            //hapus data saled yg ingin dimerge ke item tujuan
            if (saled === saled2) {
                getListDetail().remove(saled)
                break
            }
        }
        calculate()
    }

    suspend fun deleteAddon(up_saled: Saled?, saled: Saled) {
        val saledList = mutableListOf<Saled>()
        getListDetail().forEach { saled1 ->
            if (saled1 === saled) {
                saledList.add(saled1)
            }
        }
        if(saledList.size > 0){
            getListDetail().removeAll(saledList)
            calculate()
        }
        val saleAddOnDList: MutableList<SaleAddOnD> = ArrayList()
        for (saleAddOnD in addOnTrans!!.getListDetail()) {
            if (saleAddOnD.upSaled!! === up_saled && saleAddOnD.saled === saled) {
                saleAddOnDList.add(saleAddOnD)
            }
        }
        if (saleAddOnDList.size > 0) {
            deleteSaleAddonD(saleAddOnDList)
            addOnTrans!!.getListDetail().removeAll(saleAddOnDList)
        }
    }

    private fun deleteSaleAddonD(saleAddOnDList: List<SaleAddOnD>) {
//        try {
//            val saleAddOn: SaleAddOn = SaleAddOnDao.getSaleAddOnDao().readById(addOnTrans!!.getMaster())
//            val saleAddOnDS: List<SaleAddOnD> = SaleAddOnDDao.getSaleAddOnDDao().getSaleAddOnD(saleAddOn)
//            val iterator = saleAddOnDS.iterator()
//            while (iterator.hasNext()) {
//                val (id) = iterator.next()
//                for (saleAddOnd in saleAddOnDList) {
//                    if (id == saleAddOnd.id) {
////                        SaleAddOnDDao.getSaleAddOnDDao().delete(saleAddOnD);
//                    }
//                }
//            }
//        } catch (e: SQLException) {
//            e.printStackTrace()
//        }
    }

    private fun deleteOldSaleAddonD(saled: Saled) {
        val duplicateSaled: MutableList<SaleAddOnD> = ArrayList()
        for (saleAddOnD in addOnTrans!!.getListDetail()) {
            if (saled === saleAddOnD.upSaled) {
                duplicateSaled.add(saleAddOnD)
            }
        }
        addOnTrans!!.getListDetail().removeAll(duplicateSaled)
    }

    fun getSaledByUpSaledList(model : Saled)
            = addOnTrans?.let { it.getListDetail().filter { model == it.upSaled }.map { it.saled } } ?: mutableListOf()

    fun removeSalePromo(salePromo: SalePromo) {
        val salePromoRemoveList: MutableList<SalePromo> = ArrayList()
        for (salePromo1 in salePromoList) {
            if (salePromo1.promo!!.id == salePromo.promo!!.id) {
                salePromoRemoveList.add(salePromo1)
            }
            if (salePromoRemoveList.size == 2) {
                break
            }
        }
        salePromoList.removeAll(salePromoRemoveList)
    }

    fun findSalePromo(itemid: Int): SalePromo? = salePromoList.firstOrNull{ it.saled != null && it.saled!!.itemId == itemid}

    fun findSalePromo(promoType: String): SalePromo? = salePromoList.firstOrNull { it.promo!!.promoCat == promoType }

    override fun getMaster() : Sale {
        if(mTblMaster == null)
            mTblMaster = Sale()
        return mTblMaster!!
    }

}